File: python/406.py
"""Project Euler Problem 406: Guessing Game.

C(n, a, b) = minimum worst-case cost to guess a number in [1, n].
f(c) = max n guessable with budget c = 1 + f(c-a) + f(c-b).

Compute sum_{k=1}^{30} C(10^12, sqrt(k), sqrt(F_k)).
"""
from math import sqrt
import bisect

def fibonacci(k):
    """Compute k-th Fibonacci number (F_1=F_2=1)."""
    if k <= 2:
        return 1
    a, b = 1, 1
    for _ in range(k - 2):
        a, b = b, a + b
    return b

def C_func(n, a_val, b_val):
    """Compute C(n, a, b) where a and b are float values.

    Build sorted list of cost values c = i*a + j*b for non-negative i, j.
    For each c (in order), compute f(c) = 1 + f(c-a) + f(c-b).
    f(c-a) is the largest f-value for some c' <= c-a.
    Return smallest c where f(c) >= n.
    """
    # Generate all candidate cost values c = i*a + j*b up to a budget limit
    # We need f(c) >= n = 10^12. f grows roughly exponentially.
    # Max c needed is approximately log_phi(n) * max(a, b) ~ 60 * max(a, b)
    max_budget = 80.0 * max(a_val, b_val)

    # Generate sorted list of (cost, i, j) tuples
    costs = set()
    i = 0
    while i * a_val <= max_budget:
        j = 0
        while i * a_val + j * b_val <= max_budget:
            costs.add(i * a_val + j * b_val)
            j += 1
        i += 1

    sorted_costs = sorted(costs)

    # For each cost value, compute f using binary search on previous values
    # f_vals[k] = f(sorted_costs[k])
    f_vals = []
    eps = 1e-9

    for idx, c in enumerate(sorted_costs):
        if c < -eps:
            f_vals.append(0)
            continue

        # f(c-a): find largest c' <= c - a in sorted_costs
        target_a = c - a_val + eps
        pos_a = bisect.bisect_right(sorted_costs, target_a, 0, idx + 1) - 1
        fa = f_vals[pos_a] if pos_a >= 0 else 0

        # f(c-b): find largest c' <= c - b in sorted_costs
        target_b = c - b_val + eps
        pos_b = bisect.bisect_right(sorted_costs, target_b, 0, idx + 1) - 1
        fb = f_vals[pos_b] if pos_b >= 0 else 0

        f_c = 1 + fa + fb

        # Early termination check
        if f_c >= n:
            return c

        f_vals.append(f_c)

    # Should not reach here if max_budget is large enough
    return sorted_costs[-1]

def solve():
    n = 10**12
    total = 0.0

    for k in range(1, 31):
        a = sqrt(k)
        b = sqrt(fibonacci(k))
        c = C_func(n, a, b)
        total += c

    return total

if __name__ == "__main__":
    result = solve()
    print(f"{result:.8f}")

File: c/406.c
/*
 * Project Euler Problem 406: Guessing Game.
 *
 * C(n, a, b) = minimum worst-case cost to guess a number in [1, n].
 * f(c) = max n guessable with budget c = 1 + f(c-a) + f(c-b).
 *
 * Compute sum_{k=1}^{30} C(10^12, sqrt(k), sqrt(F_k)).
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef long long ll;

ll fibonacci(int k) {
    if (k <= 2) return 1;
    ll a = 1, b = 1;
    for (int i = 2; i < k; i++) {
        ll c = a + b;
        a = b;
        b = c;
    }
    return b;
}

/* Compare doubles for sorting */
int cmp_double(const void *a, const void *b) {
    double da = *(const double *)a;
    double db = *(const double *)b;
    if (da < db) return -1;
    if (da > db) return 1;
    return 0;
}

/* Binary search: find largest index where sorted_costs[idx] <= target */
int bisect_right(double *arr, int n, double target) {
    int lo = 0, hi = n;
    while (lo < hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] <= target) lo = mid + 1;
        else hi = mid;
    }
    return lo - 1;
}

double C_func(ll n, double a_val, double b_val) {
    double max_budget = 80.0 * (a_val > b_val ? a_val : b_val);

    /* Generate all candidate cost values c = i*a + j*b */
    /* Estimate max number of costs */
    int max_i = (int)(max_budget / a_val) + 2;
    int max_j = (int)(max_budget / b_val) + 2;
    int cap = (max_i + 1) * (max_j + 1);
    double *costs = (double *)malloc(cap * sizeof(double));
    int ncosts = 0;

    for (int i = 0; i * a_val <= max_budget; i++) {
        for (int j = 0; i * a_val + j * b_val <= max_budget; j++) {
            costs[ncosts++] = i * a_val + j * b_val;
        }
    }

    /* Sort and deduplicate */
    qsort(costs, ncosts, sizeof(double), cmp_double);
    int unique = 0;
    for (int i = 0; i < ncosts; i++) {
        if (i == 0 || costs[i] - costs[unique-1] > 1e-12) {
            costs[unique++] = costs[i];
        }
    }
    ncosts = unique;

    ll *f_vals = (ll *)malloc(ncosts * sizeof(ll));
    double eps = 1e-9;

    for (int idx = 0; idx < ncosts; idx++) {
        double c = costs[idx];
        if (c < -eps) {
            f_vals[idx] = 0;
            continue;
        }

        double target_a = c - a_val + eps;
        int pos_a = bisect_right(costs, idx + 1, target_a);
        ll fa = (pos_a >= 0) ? f_vals[pos_a] : 0;

        double target_b = c - b_val + eps;
        int pos_b = bisect_right(costs, idx + 1, target_b);
        ll fb = (pos_b >= 0) ? f_vals[pos_b] : 0;

        ll f_c = 1 + fa + fb;

        if (f_c >= n) {
            double result = c;
            free(costs);
            free(f_vals);
            return result;
        }

        f_vals[idx] = f_c;
    }

    double result = costs[ncosts - 1];
    free(costs);
    free(f_vals);
    return result;
}

int main(void) {
    ll n = 1000000000000LL;  /* 10^12 */
    double total = 0.0;

    for (int k = 1; k <= 30; k++) {
        double a = sqrt((double)k);
        double b = sqrt((double)fibonacci(k));
        double c = C_func(n, a, b);
        total += c;
    }

    printf("%.8f\n", total);
    return 0;
}
